site/_posts/2019-11-18-cis-benchmark-plugin.md

```diff
@@ -0,0 +1,186 @@
+- Modify the plugin definition with our custom `podSpec` options
+
+First, we need to decide on the command that the plugin should run. The `kube-bench` image defaults to printing results to stdout. Instead, we want to redirect them to a file (kube-bench has a flag for this) and then notify Sonobuoy when complete. 
+
+We can chain these desired commands together as a single bash command and then save the plugin definition to a file:
+
+```
+$ sonobuoy gen plugin \
+    --name kube-bench-worker \
+    --image=schnake/kube-bench:v0.2.0-demo \
+    --cmd=/bin/sh
+    --args=-c
+    --args=”kube-bench --version 1.13 --outputfile /tmp/results/output.xml --junit ; echo -n /tmp/results/output.xml > /tmp/results/done” \
+    --show-default-podspec > kube-bench-worker.yaml
+```
+
+Now open up `kube-bench-worker.yaml` and modify the `podSpec` to include the extra required volumes and `hostPID: true`:
+
+```
+podSpec:
+  containers: []
+  dnsPolicy: ClusterFirstWithHostNet
+  hostIPC: true
+  hostNetwork: true
+  hostPID: true
+  serviceAccountName: sonobuoy-serviceaccount
+  tolerations:
+  - operator: Exists
+  volumes:
+  - name: var-lib-etcd
+    hostPath:
+      path: "/var/lib/etcd"
+  - name: var-lib-kubelet
+    hostPath:
+      path: "/var/lib/kubelet"
+  - name: etc-systemd
+    hostPath:
+      path: "/etc/systemd"
+  - name: etc-kubernetes
+    hostPath:
+      path: "/etc/kubernetes"
+  - name: usr-bin
+    hostPath:
+      path: "/usr/bin"
+```
+
+_Note: All these values came from reading the [sample code](https://github.com/aquasecurity/kube-bench/blob/master/job.yaml) in the kube-bench repo._
+
+Lastly, we need to understand that the CIS benchmarks come in two different flavors: one for worker nodes and one for master nodes. Therefore, we need to run different versions of the plugin on the different nodes. To accomplish this, we will:
+
+- Add the appropriate [node affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) to the worker node plugin so it runs on the right nodes.
+- Copy the plugin definition to a new file and transform the necessary values for the master plugin name, command, and node affinity.
+
+The affinity section for the worker node plugin should look like this:
+
+```
+affinity:
+    nodeAffinity: 
+      requiredDuringSchedulingIgnoredDuringExecution: 
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: node-role.kubernetes.io/master
+            operator: DoesNotExist
+```
+
+For the master nodes, we need to do three things:
+
+- Change `DoesNotExist` to `Exists` to target the correct nodes
+- Change the plugin name so the plugins have unique names, which is a Sonobuoy requirement
+- Tweak the containers command so that the correct set of checks is run
+
+Each of these things can be done manually, or with the following command:
+
+```
+$ cat kub-bench-worker.yaml | \
+  sed ‘/kube-bench-worker/kube-bench-master/’ | \
```

Comment:
Hmm, I tried this out but these `sed` commands don't work for me. I think they should be:
```suggestion
  sed 's/kube-bench-worker/kube-bench-master/g' | \
```